experimental/casemapping/src/internals.rs

```diff
@@ -0,0 +1,1113 @@
+                return FullMappingResult::CodePoint(mapped);
+            }
+            if kind == MappingKind::Fold && self.exceptions.no_simple_case_folding(idx) {
+                return FullMappingResult::CodePoint(c);
+            }
+
+            if let Some(slot_char) = self.exceptions.slot_char_for_kind(idx, kind) {
+                FullMappingResult::CodePoint(slot_char)
+            } else {
+                FullMappingResult::CodePoint(c)
+            }
+        }
+    }
+
+    pub(crate) fn to_full_lower(
+        &self,
+        c: char,
+        context: ContextIterator,
+        locale: CaseMapLocale,
+    ) -> FullMappingResult {
+        self.full_helper(c, context, locale, MappingKind::Lower)
+    }
+
+    pub(crate) fn to_full_upper(
+        &self,
+        c: char,
+        context: ContextIterator,
+        locale: CaseMapLocale,
+    ) -> FullMappingResult {
+        self.full_helper(c, context, locale, MappingKind::Upper)
+    }
+
+    // Titlecase mapping is not yet exposed
+    #[allow(dead_code)]
+    pub(crate) fn to_full_title(
+        &self,
+        c: char,
+        context: ContextIterator,
+        locale: CaseMapLocale,
+    ) -> FullMappingResult {
+        self.full_helper(c, context, locale, MappingKind::Title)
+    }
+
+    // Note: in ICU4C, case folding takes an options bag instead of a locale,
+    // with the only defined option being whether or not to use Turkic (T)
+    // mappings for dotted/dotless i. In ICU4X, we expose a similar locale-free
+    // API for case folding, but internally represent this as a Turkish locale
+    // to simplify shared code.
+    pub(crate) fn to_full_folding(
+        &self,
+        c: char,
+        context: ContextIterator,
+        locale: CaseMapLocale,
+    ) -> FullMappingResult {
+        self.full_helper(c, context, locale, MappingKind::Fold)
+    }
+
+    // These constants are used for hardcoded locale-specific foldings.
+    const I_DOT: &'static str = "\u{69}\u{307}";
+    const J_DOT: &'static str = "\u{6a}\u{307}";
+    const I_OGONEK_DOT: &'static str = "\u{12f}\u{307}";
+    const I_DOT_GRAVE: &'static str = "\u{69}\u{307}\u{300}";
+    const I_DOT_ACUTE: &'static str = "\u{69}\u{307}\u{301}";
+    const I_DOT_TILDE: &'static str = "\u{69}\u{307}\u{303}";
+
+    // Special case folding mappings, hardcoded.
+    // This handles the special Turkic mappings for uppercase I and dotted uppercase I
+    // For non-Turkic languages, this mapping is normally not used.
+    // For Turkic languages (tr, az), this mapping can be used instead of the normal mapping for these characters.
+    fn simple_fold_special_case(&self, c: char, options: FoldOptions) -> char {
+        debug_assert!(c == '\u{49}' || c == '\u{130}');
+        let is_turkic = options.exclude_special_i;
+        match (c, is_turkic) {
+            // Turkic mappings
+            ('\u{49}', true) => '\u{131}', // 0049; T; 0131; # LATIN CAPITAL LETTER I
```

Comment:
suggestion: have named constants for chars, too

Could these chars have named consatants like lines 623-628 (more readable)?

Reply:
I mostly followed the ICU4C code here. I can take another pass and clean it up (in a future patch).